src/pow/randomx.rs

```diff
@@ -0,0 +1,135 @@
+pub use randomx_rs::RandomXFlag;
+use randomx_rs::{RandomXCache, RandomXDataset, RandomXError, RandomXVM};
+use rayon::prelude::{IntoParallelIterator, ParallelIterator};
+use thread_local::ThreadLocal;
+
+use super::Error;
+
+const RANDOMX_CACHE_KEY: &[u8] = b"spacemesh-randomx-cache-key";
+
+impl From<randomx_rs::RandomXError> for Error {
+    fn from(e: randomx_rs::RandomXError) -> Self {
+        Error::Internal(Box::new(e))
+    }
+}
+
+pub struct PoW {
+    cache: Option<RandomXCache>,
+    dataset: Option<RandomXDataset>,
+    flags: RandomXFlag,
+    vms: ThreadLocal<RandomXVM>,
+}
+
+impl PoW {
+    pub fn new(flags: RandomXFlag) -> Result<PoW, Error> {
+        let cache = RandomXCache::new(flags, RANDOMX_CACHE_KEY)?;
+        let (cache, dataset) = if flags.contains(RandomXFlag::FLAG_FULL_MEM) {
+            (None, Some(RandomXDataset::new(flags, cache, 0)?))
+        } else {
+            (Some(cache), None)
+        };
+
+        Ok(Self {
+            cache,
+            dataset,
+            flags,
+            vms: ThreadLocal::new(),
+        })
+    }
+
+    fn get_vm(&self) -> Result<&RandomXVM, RandomXError> {
+        self.vms
+            .get_or_try(|| RandomXVM::new(self.flags, self.cache.clone(), self.dataset.clone()))
+    }
+
+    pub fn prove(
+        &self,
+        nonce_group: u8,
+        challenge: &[u8; 8],
+        difficulty: &[u8; 32],
+    ) -> Result<u64, Error> {
+        let pow_input = [[0u8; 7].as_slice(), [nonce_group].as_slice(), challenge].concat();
+
+        let (pow_nonce, _) = (0..2u64.pow(56))
+            .into_par_iter()
+            .map_init(
+                || -> Result<_, Error> { Ok((self.get_vm()?, pow_input.clone())) },
```

Comment:
so prover reuses same cache for parallel map? i am confused why we are not sharing it with verifier